fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ForceCTEReuseRule.java

```diff
@@ -0,0 +1,212 @@
+        public static boolean hasNonDeterministicFunction(OptExpression root) {
+            return new NonDeterministicVisitor().visit(root, null);
+        }
+
+        boolean checkColumnRefMap(Map<ColumnRefOperator, ScalarOperator> columnRefMap) {
+            if (columnRefMap == null) {
+                return false;
+            }
+            for (ScalarOperator ref : columnRefMap.values()) {
+                if (hasNonDeterministicFunc(ref)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        private boolean hasNonDeterministicFunc(ScalarOperator scalarOperator) {
+            if (scalarOperator instanceof CallOperator) {
+                String fnName = ((CallOperator) scalarOperator).getFnName();
+                if (FunctionSet.nonDeterministicFunctions.contains(fnName)) {
+                    return true;
+                }
+            } else if (scalarOperator instanceof LambdaFunctionOperator) {
+                LambdaFunctionOperator lambdaOp = (LambdaFunctionOperator) scalarOperator;
+                Map<ColumnRefOperator, ScalarOperator> columnRefMap = lambdaOp.getColumnRefMap();
+                if (checkColumnRefMap(columnRefMap)) {
+                    return true;
+                }
+            }
+            for (ScalarOperator child : scalarOperator.getChildren()) {
+                if (hasNonDeterministicFunc(child)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        private boolean checkAggCall(Map<ColumnRefOperator, CallOperator> aggregations) {
+            for (Map.Entry<ColumnRefOperator, CallOperator> entry : aggregations.entrySet()) {
+                CallOperator aggCall = entry.getValue();
+                for (ScalarOperator arg : aggCall.getArguments()) {
+                    if (hasNonDeterministicFunc(arg)) {
+                        return true;
+                    }
+                }
+            }
+            return false;
+        }
+
+        private boolean checkProject(Projection projection) {
+            if (projection == null) {
+                return false;
+            }
+            Map<ColumnRefOperator, ScalarOperator> columnRefMap =
+                    projection.getColumnRefMap();
+            if (columnRefMap == null) {
+                return false;
+            }
+            for (ScalarOperator scalarOperator : columnRefMap.values()) {
+                if (hasNonDeterministicFunc(scalarOperator)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        private boolean checkOptExpression(OptExpression optExpression) {
+            Operator operator = optExpression.getOp();
+            if (operator.getProjection() != null && checkProject(operator.getProjection())) {
+                return true;
+            }
```

Comment:
Predicate can be pushed down to each operator. check predicate process may like projection.